src/lib/updating-element.ts

```diff
@@ -0,0 +1,503 @@
+   * If true, then if present the `type.toAttribute` function is used to serialize
+   * the value; otherwise, the property value is used.
+   */
+  reflect?: boolean;
+  /**
+   * Describes if setting a property should trigger invalidation and updating.
+   * This function takes the new and oldValue and returns true if invalidation
+   * should occur. If not present, a strict identity check is used.
+   */
+  shouldInvalidate?(value: unknown, oldValue: unknown): boolean;
+}
+
+/**
+ * Object that describes accessors to be created on the element prototype.
+ * An accessor is created for each key with the given options.
+ */
+export interface Properties {
+  [key: string]: PropertyOptions;
+}
+
+interface AttributeMap {
+  [key: string]: string;
+}
+
+interface AnyObject {
+  [key: string]: {};
+}
+
+/**
+ * Creates a property accessor on the given prototype if one does not exist.
+ * Uses `getProperty` and `setProperty` to manage the property's value.
+ */
+const makeProperty = (name: string, proto: Object) => {
+  if (proto.hasOwnProperty(name) || (name in proto)) {
+    return;
+  }
+  Object.defineProperty(proto, name, {
+    get() {
+      return this.getProperty(name);
+    },
+    set(value) {
+      this.setProperty(name, value);
+    },
+    configurable: true,
+    enumerable: true
+  });
+};
+
+/**
+ * Creates and sets object used to memoize all class property values. Object
+ * is chained from superclass.
+ */
+const ensurePropertyStorage = (ctor: typeof UpdatingElement) => {
+  if (!ctor.hasOwnProperty('_classProperties')) {
+    ctor._classProperties = Object.create(Object.getPrototypeOf(ctor)._classProperties);
+  }
+};
+
+/**
+ * Decorator which creates a property. Optionally a `PropertyOptions` object
+ * can be supplied to describe how the property should be configured.
+ */
+export const property = (options?: PropertyOptions) => (proto: Object, name: string) => {
+  const ctor = proto.constructor as typeof UpdatingElement;
+  ensurePropertyStorage(ctor);
+  ctor._classProperties[name] = options || {};
+  makeProperty(name, proto);
+};
+
+
+/**
+ * AttributeProcessor which configures properties which should reflect to and
+ * from boolean attributes. If the attribute exists, the property becomes true;
+ * and if the property value is truthy, the attribute is set to an empty string.
+ */
+export const BooleanAttribute: AttributeProcessor = {
+  fromAttribute: (value: string) => value !== null,
+  toAttribute: (value: string) => value ? '' : null
+};
+
+/**
+ * Change function that returns true if `value` is different from `oldValue`.
+ * This method is used as the default for a property's `shouldChange` function.
+ */
+// tslint:disable-next-line no-any
+export const identity = (value: any, old: any) => {
+  // This ensures (old==NaN, value==NaN) always returns false
+  return old !== value && (old === old || value === value);
+};
+
+enum ValidationState {
```

Comment:
I _really_ prefer avoiding enums and namespaces in TypeScript. They're the only non-type extensions to JS, and I think generally considered a mistake. Unless you need the bidirectional mapping of enums, using a plain object or a union type of literals usually suffices, ie:

```ts
type ValidationState = 'disabled' | 'valid' | 'invalid';
```

or:

```ts
const disabled = 0;
const valid = 1;
const invalid = 2;
type ValidationState = typeof disabled | typeof valid | typeof invalid;
```

Reply:
Done